### PR TITLE
8241502: C2: Migrate x86_64.ad to MacroAssembler

### PR DESCRIPTION
**Pull request by JohnTortugo**
Relates to: https://bugs.openjdk.java.net/browse/JDK-8241502
Tested on: Linux tier1, 2 and 3

Can you please take a look whether these changes are going in the direction expected or not? If it is, I'll continue working on the `JDK-8241502` but I'd like to split it in a few PRs since it's a lot of changes.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8241502](https://bugs.openjdk.java.net/browse/JDK-8241502): C2: Migrate x86_64.ad to MacroAssembler

### Reviewers
 * [Vladimir Ivanov](https://openjdk.java.net/census#vlivanov) (@iwanowww - **Reviewer**) ⚠️ Review applies to ef7a993ca5e012813b9c2dd96ba91859970f5f62
 * [Vladimir Kozlov](https://openjdk.java.net/census#kvn) (@vnkozlov - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/2420/head:pull/2420` \
`$ git checkout pull/2420`

Update a local copy of the PR: \
`$ git checkout pull/2420` \
`$ git pull https://git.openjdk.java.net/jdk pull/2420/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 2420`

View PR using the GUI difftool: \
`$ git pr show -t 2420`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/2420.diff">https://git.openjdk.java.net/jdk/pull/2420.diff</a>

</details>

### CODE REVIEW

**Comment by iwanowww** [comment]
> Can you please take a look whether these changes are going in the direction expected or not?

Yes, the patch is perfectly aligned with what JDK-8241502 proposes. Thanks for taking care of it.

**Comment by vnkozlov** [comment]
I agree. We wanted to do that for long time.

**Comment by iklam** [comment]
I am curious if the x86_64.o file changes in any significant way (speed of size).

**Comment by dean-long** [comment]
I wish there was a way for the old and new versions to co-exist at the same time, so we could generate the code the old way and and the new way, then compare, for automatic verification of the MacroAssember version.

**Comment by JohnTortugo** [comment]
Thank you all for the feedback!

@iklam - I'll check that and let you know once I make more conversions.

@dean-long - That would be great. I'm all ears for the best way to test this!

**Comment by dean-long** [comment]
Here's one way to test both versions, using loadRange as an example.  It's not exactly pretty, but it seems to work.
[loadRange.txt](https://github.com/openjdk/jdk/files/5994725/loadRange.txt)

**Comment by JohnTortugo** [comment]
@dean-long - I recently hacked something pretty similar - basically, I added a flag to the CodeSection class to make it print emitted bytes whenever the flag was set, then I created two encoding classes in the AD file to toggle the print flag. That way I was able to _visually_ compare the output of the two versions of the code. Your approach with the CRC is much better. Thanks a lot!

**Comment by JohnTortugo** [comment]
Hi again, here is a quick update on the progress of this and an ask for advice.

I finished doing all conversions. I'm using an automated approach to validate the conversions: for each instruction that I converted I created an `enc_class` (just for debugging purposes) with pretty much the code that @dean-long provided above. However, I'm hitting a limitation on test coverage. I performed 248 conversions, but until now I was able to test only 203! I've tested tier1, tier2, and tier3 with different GCs, different VM options, on Linux and macOS. In other words, I've tried several things that I could think of but I couldn't make C2 use some instructions. 

I created a [Gist with the list of instructions I couldn't test so far](https://gist.github.com/JohnTortugo/c6d09a8ef6160422d6bf0ae5f894c1fe). Can you suggest something I can do to make C2 use those additional instructions?

**Comment by dean-long** [comment]
Was -Xcomp one of the flags you tried?  If so, then you may need to write new tests for those instructions without test coverage.  Or you could do the 203 first and file a new RFE for the remaining.

**Comment by JohnTortugo** [comment]
Hi @dean-long, thanks for the suggestions. Yes, "-Xcomp" is among the flags that I tried. In the end, what helped the most was commenting out some optimizations that simplify the IR graph - i.e., I commented out a bunch of <IRNode>::Ideal methods. I also tried commenting out some instructions in the .ad file and also changing `ins_costs`. These things helped quite a lot and I was able to get 26 more conversions validated; now the total of conversions + validations is 239. The conversions that I couldn't validate I didn't include in the PR; I'll follow your suggestion and create an RFE once the current PR is merged.

So, please consider this PR ready for review. I've tested these changes on Linux, Windows, and macOS with fastdebug and release binaries running jdk_tier1/2/3 and hotspot_all. Tests were also performed with different GCs: ZGC, Shenandoah, and G1.

I didn't remove the, now unused, `enc_classes` and related code; I think it would be better to keep them for a few more days to make easier any eventual debugging. I'll be happy to remove them if you think that's a better approach.

**Comment by dean-long** [comment]
Hi @JohnTortugo, good job on the instruction validation.  I'm OK with leaving the unused enc_classes for now, but let's see what the other reviewers think.  One option would be to move them to a separate .ad file.

I guess you counted an instruction validated if it was generated at least once (so at least 1 possible encoding)?  On RISC architectures that would probably be enough, but I know on x86/x64, register choice can change the instruction encoding and length, thanks to prefixes, etc.  Unfortunately, I'm out of clever ideas on how to verify _all_ possible encodings of the new AD instructions.

**Review by iwanowww — APPROVED** [review]
Looks good.

**Comment by iwanowww** [comment]
> I didn't remove the, now unused, enc_classes and related code; I think it would be better to keep them for a few more days to make easier any eventual debugging. I'll be happy to remove them if you think that's a better approach.

Can you elaborate, please, what you mean here? I don't see "enc_classes and related code" in the latest patch.

**Comment by iwanowww** [comment]
> I guess you counted an instruction validated if it was generated at least once (so at least 1 possible encoding)? On RISC architectures that would probably be enough, but I know on x86/x64, register choice can change the instruction encoding and length, thanks to prefixes, etc.

One possibility is to write a native test (using gtest) and exercise some encoding methods by enumerating all the possible input value locations. Probably, too much for a regularly executed test (there's a lot of stuff missing right now to support such scenario), but something which can be quickly hacked for a one-time run.

**Review by vnkozlov — APPROVED** [review]

**Comment by vnkozlov** [comment]
> > I guess you counted an instruction validated if it was generated at least once (so at least 1 possible encoding)? On RISC architectures that would probably be enough, but I know on x86/x64, register choice can change the instruction encoding and length, thanks to prefixes, etc.
> 
> One possibility is to write a native test (using gtest) and exercise some encoding methods by enumerating all the possible input value locations. Probably, too much for a regularly executed test (there's a lot of stuff missing right now to support such scenario), but something which can be quickly hacked for a one-time run.

I suggest to do that as separate RFE. This changes could be tested by running in all our tiers testing.

Cleaning unused code could be also done separately.

**Comment by dean-long** [comment]
> Can you elaborate, please, what you mean here? I don't see "enc_classes and related code" in the latest patch.

I was confused by this at first too.  I believe the question is, should unused code such as "enc_class cdql_enc" be removed as part of this PR?

**Comment by JohnTortugo** [comment]
Thank you all for reviewing!

> Can you elaborate, please, what you mean here? I don't see "enc_classes and related code" in the latest patch.

@dean-long got my wording right. I'll do the cleaning as a separate RFE.

> This changes could be tested by running in all our tiers testing.

Right, this is how I tested: hotspot_all, tier1, tier2, tier3, different GCs, different OS's.

**Comment by iwanowww** [comment]
>  I'll do the cleaning as a separate RFE.

Sounds good.

File: src/hotspot/cpu/x86/x86_64.ad [hunk 1]
```diff
@@ -8325,8 +8388,19 @@ instruct divI_rReg(rax_RegI rax, rdx_RegI rdx, no_rax_rdx_RegI div,
     "normal: cdql\n\t"
             "idivl   $div\n"
     "done:"        %}
-  opcode(0xF7, 0x7);  /* Opcode F7 /7 */
-  ins_encode(cdql_enc(div), REX_reg(div), OpcP, reg_opc(div));
```
Comment (iwanowww) [inline src/hotspot/cpu/x86/x86_64.ad:8329]: `cdql_enc`/`cdqq_enc` contain extensive comments about the logic. What about migrating them to dedicated methods in `MacroAssembler`/`C2_MacroAssembler` (`idivrem`/`ldivrem`) and calling those from `divI_rReg` / `divL_rReg`/`divModL_rReg_divmod`/...?

Comment (JohnTortugo) [reply src/hotspot/cpu/x86/x86_64.ad:8329]: Hi Vladimir, thanks for the comment. Actually, there is already an implementation of cdql/cdqq in MacroAssembler, the names are slightly different though: corrected_idivl and corrected_idivq. I followed your suggestion and replaced the encoding of div*_rReg / divMod*_rReg_divmod and mod*_rReg with calls to these two functions.

**Comment by JohnTortugo** [comment]
@iwanowww @vnkozlov - can you please take a look again? I think I've addressed all your comments. Thank you.

**Review by iwanowww — APPROVED** [review]
Looks good.

**Review by vnkozlov — APPROVED** [review]

**Comment by vnkozlov** [comment]
I will sponsor it after my testing is passed.

**Comment by vnkozlov** [comment]
Build failed on Windows:
```
t:\workspace\open\src\hotspot\cpu\x86\assembler_x86.cpp(10877): error C2220: the following warning is treated as an error
t:\workspace\open\src\hotspot\cpu\x86\assembler_x86.cpp(10877): warning C4309: 'argument': truncation of constant value
```

**Review by vnkozlov — CHANGES_REQUESTED** [review]
Please, fix build failure and run testing on Windows.

**Comment by JohnTortugo** [comment]
Thank you, Vladimir. I'll take a look on that!

File: src/hotspot/cpu/x86/assembler_x86.cpp [hunk 2]
```diff
@@ -9234,14 +9410,14 @@ void Assembler::pusha() { // 32bit
 }
 
 void Assembler::set_byte_if_not_zero(Register dst) {
-  emit_int24(0x0F, (unsigned char)0x95, (0xE0 | dst->encoding()));
+  emit_int24(0x0F, (unsigned char)0x95, (0xC0 | dst->encoding()));
```
Comment (vnkozlov) [inline src/hotspot/cpu/x86/assembler_x86.cpp:9413]: Can you explain this change?
I see it is used only in c1_LIRAssembler_x86.cpp

Comment (JohnTortugo) [reply src/hotspot/cpu/x86/assembler_x86.cpp:9413]: The operand encoding of SETNE r/m8 is just 0b11.000.000 AFAIU, and not 0b11.100.000 as in the previous code. I found this difference while comparing the bytes emitted by instructions encoded using "opcode(...)" with instructions emitted by MacroAssembler.

Comment (JohnTortugo) [reply src/hotspot/cpu/x86/assembler_x86.cpp:9413]: I.e., I basically patched the MacroAssembler routine to emit an instruction that matched this: https://github.com/openjdk/jdk/blob/b84f6901275aedfeb59438898846b6e69ced316d/src/hotspot/cpu/x86/x86_64.ad#L2750

Comment (vnkozlov) [reply src/hotspot/cpu/x86/assembler_x86.cpp:9413]: Got it. I submitted new testing.

**Comment by vnkozlov** [comment]
I got a lot of failures on Windows only. Almost all test/hotspot/jtreg/serviceability/sa/ tests failed and others.
Examples:
```
serviceability/sa/TestSysProps.java

#  Internal Error (t:\\workspace\\open\\src\\hotspot\\share\\runtime\\stackValue.cpp:139), pid=7572, tid=26052
#  assert(oopDesc::is_oop_or_null(val, false)) failed: bad oop found

compiler/c2/Test6800154.java
java.lang.InternalError: 9223372036854775807 / -9223372036854775808 failed: 65547 != 0
	at compiler.c2.Test6800154.run(Test6800154.java:111)
	at compiler.c2.Test6800154.main(Test6800154.java:97)
```

**Comment by JohnTortugo** [comment]
@vnkozlov - thank you so much for running the tests! The cause of the problems you reported were the last changes I made to the div/mod instructions. I fixed the code and ran all tests again on Linux, macOS, and Windows and they are looking good (jdk tier1, 2, 3, and hotspot_all.

**Comment by vnkozlov** [comment]
I started new testing.

**Comment by vnkozlov** [comment]
Preliminary result shows compiler/c2/cr6340864/TestLongVect.java test failure on linux and windows -x64 when run with not G1 gc (Serial, Parallel, ZGC) (the line could be different again due to use of latest source base):

#  Internal Error (/workspace/open/src/hotspot/cpu/x86/assembler_x86.cpp:10753), pid=9483, tid=9495
#  assert(isByte(imm8)) failed: not a byte
#
# JRE version: Java(TM) SE Runtime Environment (17.0) (fastdebug build 17-internal+0-LTS-2021-04-30-0438565.jdkgit)
# Java VM: Java HotSpot(TM) 64-Bit Server VM (fastdebug 17-internal+0-LTS-2021-04-30-0438565.jdkgit, mixed mode, sharing, tiered, compressed oops, compressed class ptrs, parallel gc, linux-amd64)
# Problematic frame:
# V  [libjvm.so+0x6471b4]  Assembler::sarq(RegisterImpl*, int)+0x164

V  [libjvm.so+0x6471b4]  Assembler::sarq(RegisterImpl*, int)+0x164
V  [libjvm.so+0x34d566]  sarL_rReg_immNode::emit(CodeBuffer&, PhaseRegAlloc*) const+0xb6
V  [libjvm.so+0x1588551]  PhaseOutput::scratch_emit_size(Node const*)+0x421
V  [libjvm.so+0x157f3f8]  PhaseOutput::shorten_branches(unsigned int*)+0x2b8
V  [libjvm.so+0x159165a]  PhaseOutput::Output()+0xcca
V  [libjvm.so+0xa0636b]  Compile::Code_Gen()+0x43b

**Comment by vnkozlov** [comment]
@JohnTortugo  did you fix the last issue? Let me know when I should test it.

**Comment by JohnTortugo** [comment]
@vnkozlov - I've fixed the code, I'm finishing running some more tests. I'll push the new changes later today. Thanks!

**Comment by JohnTortugo** [comment]
Hi @vnkozlov - I just pushed a patch that will handle the assert triggering problem. Thanks again for helping!

**Comment by vnkozlov** [comment]
I started testing for latest changes (10).

**Review by vnkozlov — APPROVED** [review]
tier1-5 passed clean

**Comment by JohnTortugo** [comment]
Thank you @vnkozlov !

/integrate

**Pull request closed without merging**